jax/experimental/jax2tf/jax2tf.py

```diff
@@ -87,13 +88,15 @@ def _safe_convert_to_tensor(val, dtype=None) -> TfVal:
   `ndarray` with a `uint16` dtype, for which conversion is properly defined. Then, we
   simply bitcast it back to `bfloat16`.
   """
-  dtype = dtype if dtype else (val.dtype if hasattr(val, "dtype") else None)
-  if (dtype == jnp.bfloat16 or isinstance(val, jnp.bfloat16)):
-    if not isinstance(val, jnp.ndarray):
-      val = np.array(val, jnp.bfloat16)
-
-    val = tf.bitcast(tf.convert_to_tensor(val.view(jnp.uint16),
-                                          dtype=to_tf_dtype(jnp.uint16)),
+  dtype = dtype or _dtype(val)
+
+  if (dtype == jnp.bfloat16 or _dtype(val) == jnp.bfloat16):
+    # TODO: handle bfloat16 without going through numpy.
+    # We cannot call val.view() on a DeviceArray, because it passes val to
+    # bitcast_convert_type_p, which calls back here, leading to an infinite
+    # recursion.  Instead, copy to a numpy array and create the view there.
+    val = np.asarray(val, jnp.bfloat16).view(np.uint16)
```

Comment:
Since the JAX and TF `bfloat16` actually have the same representation, even though they look like different types, is it possible to do the conversion without any array copy?

Reply:
@bchetioui PTAL

Reply:
I can not think of a way to do it from inside jax2tf. However, with a little more careful of handling of dtypes (i.e. adding calls to `to_jax_dtype` and `to_tf_dtype` where it is necessary to get the expected bfloat16 representation), we should be able to call to make a call like `jax2tf.convert(func_jax)(jnp_bfloat16_array.view(tf.bfloat16.as_numpy_dtype))`.

This is not ideal – as it puts a burden on the user, and adds some clutter to the code (which seems like it might be annoying to test thoroughly). The ideal solution would be for the TF and JAX bfloat16 to be merged together, and then all this logic can go away.

Reply:
This merging of bfloat16 is now in progress, perhaps we can wait with this PR until that merging lands in TF.

Reply:
The bug this PR addresses is blocking work on weak types, so I’d prefer not to delay the fix.

Reply:
Another option would be to wait, but disable the problematic tests, with the understanding that changes to `lax.convert_element_type` will break some uses of `bfloat16` in jax2tf.